File: 贪心/135分发糖果/135.分发糖果.java
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=135 lang=java
 *
 * [135] 分发糖果
 */

// @lc code=start
class Solution {
    public int candy(int[] ratings) {
        int[] left = new int[ratings.length];
        Arrays.fill(left, 1);

        for(int i = 1; i < ratings.length; i++) {
            if(ratings[i] > ratings[i-1]) {
                left[i] = left[i-1] +1;
            }
        }

        int right = 0;
        for(int i = ratings.length-1; i >= 0 ; i--) {
            if(i < ratings.length-1 && ratings[i] > ratings[i+1]) {
                left[i] = Math.max(left[i], left[i+1]+1);
            }
            right += left[i];
        }

        return right;
    }
}
// @lc code=end

File: 贪心/135分发糖果/135.分发糖果.js
/*
 * @lc app=leetcode.cn id=135 lang=javascript
 *
 * [135] 分发糖果
 */

// @lc code=start
/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function(ratings) {
  let arr = new Array(ratings.length).fill(1);
    for(let i = 1; i < ratings.length; i++) {
        if(ratings[i] > ratings[i-1]) {
            arr[i] = arr[i-1] + 1;
        }
    }

    for(let i = ratings.length-2 ; i >=0 ; i-- ) {
        if(ratings[i] > ratings[i+1]) {
            arr[i] = Math.max(arr[i],arr[i+1]+1);
        }
    }
    return arr.reduce((cur,pre)=>{
        return cur+pre;
    },0);
};
// @lc code=end

